pkg/pool-gyro/contracts/lib/GyroECLPMath.sol

```diff
@@ -0,0 +1,777 @@
+// SPDX-License-Identifier: LicenseRef-Gyro-1.0
+// for information on licensing please see the README in the GitHub repository
+// <https://github.com/gyrostable/concentrated-lps>.
+
+pragma solidity ^0.8.24;
+
+import { SafeCast } from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+import { IGyroECLPPool } from "@balancer-labs/v3-interfaces/contracts/pool-gyro/IGyroECLPPool.sol";
+
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+
+import { SignedFixedPoint } from "./SignedFixedPoint.sol";
+import { GyroPoolMath } from "./GyroPoolMath.sol";
+
+/**
+ * @notice ECLP math library. Pretty much a direct translation of the python version.
+ * @dev We use *signed* values here because some of the intermediate results can be negative (e.g. coordinates of
+ * points in the untransformed circle, "prices" in the untransformed circle).
+ */
+library GyroECLPMath {
+    using SignedFixedPoint for int256;
+    using FixedPoint for uint256;
+    using SafeCast for uint256;
+    using SafeCast for int256;
+
+    error RotationVectorWrong();
+    error RotationVectorNotNormalized();
+    error AssetBoundsExceeded();
+    error DerivedTauNotNormalized();
+    error StretchingFactorWrong();
+    error DerivedUvwzWrong();
+    error InvariantDenominatorWrong();
+    error MaxAssetsExceeded();
+    error MaxInvariantExceeded();
+    error DerivedDsqWrong();
+
+    uint256 internal constant _ONEHALF = 0.5e18;
+    int256 internal constant _ONE = 1e18; // 18 decimal places
+    int256 internal constant _ONE_XP = 1e38; // 38 decimal places
+
+    // Anti-overflow limits: Params and DerivedParams (static, only needs to be checked on pool creation).
+    int256 internal constant _ROTATION_VECTOR_NORM_ACCURACY = 1e3; // 1e-15 in normal precision
+    int256 internal constant _MAX_STRETCH_FACTOR = 1e26; // 1e8   in normal precision
+    int256 internal constant _DERIVED_TAU_NORM_ACCURACY_XP = 1e23; // 1e-15 in extra precision
+    int256 internal constant _MAX_INV_INVARIANT_DENOMINATOR_XP = 1e43; // 1e5   in extra precision
+    int256 internal constant _DERIVED_DSQ_NORM_ACCURACY_XP = 1e23; // 1e-15 in extra precision
+
+    // Anti-overflow limits: Dynamic values (checked before operations that use them).
+    int256 internal constant _MAX_BALANCES = 1e34; // 1e16 in normal precision
+    int256 internal constant _MAX_INVARIANT = 3e37; // 3e19 in normal precision
+
+    struct QParams {
+        int256 a;
+        int256 b;
+        int256 c;
+    }
+
+    /// @dev Enforces limits and approximate normalization of the rotation vector.
+    function validateParams(IGyroECLPPool.EclpParams memory params) internal pure {
+        if (0 > params.s || params.s > _ONE) {
+            revert RotationVectorWrong();
+        }
+
+        if (0 > params.c || params.c > _ONE) {
+            revert RotationVectorWrong();
+        }
```

Comment:
This makes me want to use `require` :face_with_spiral_eyes: logic looks good though.
We can probably use slightly different errors here for C and S; that can't hurt.

Reply:
Hm. All of this is fine, but I'm wondering whether readability would be better doing `if (!original_logic) { revert }` instead of denying the original logic manually.

Another option is to compile with 0.8.27 and use require. Maybe we can just compile this package with 0.8.27?

Reply:
Done!